receiver/libhoneyreceiver/receiver.go

```diff
@@ -0,0 +1,241 @@
+	go func() {
+		defer r.shutdownWG.Done()
+
+		if err := r.server.Serve(hln); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			componentstatus.ReportStatus(host, componentstatus.NewFatalErrorEvent(err))
+		}
+	}()
+	return nil
+}
+
+func (r *libhoneyReceiver) Start(ctx context.Context, host component.Host) error {
+	if err := r.startHTTPServer(ctx, host); err != nil {
+		return errors.Join(err, r.Shutdown(ctx))
+	}
+
+	return nil
+}
+
+// Shutdown is a method to turn off receiving.
+func (r *libhoneyReceiver) Shutdown(ctx context.Context) error {
+	var err error
+
+	if r.server != nil {
+		err = r.server.Shutdown(ctx)
+	}
+
+	r.shutdownWG.Wait()
+	return err
+}
+
+func (r *libhoneyReceiver) registerTraceConsumer(tc consumer.Traces) {
+	r.nextTraces = tc
+}
+
+func (r *libhoneyReceiver) registerLogConsumer(tc consumer.Logs) {
+	r.nextLogs = tc
+}
+
+func (r *libhoneyReceiver) handleSomething(resp http.ResponseWriter, req *http.Request) {
+	enc, ok := readContentType(resp, req)
+	if !ok {
+		return
+	}
+
+	dataset, err := getDatasetFromRequest(req.RequestURI)
+	if err != nil {
+		r.settings.Logger.Info("No dataset found in URL", zap.String("req.RequstURI", req.RequestURI))
+	}
+
+	for _, p := range r.cfg.HTTP.TracesURLPaths {
+		dataset = strings.Replace(dataset, p, "", 1)
+		r.settings.Logger.Debug("dataset parsed", zap.String("dataset.parsed", dataset))
+	}
+
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		errorutil.HTTPError(resp, err)
+	}
+	if err = req.Body.Close(); err != nil {
+		errorutil.HTTPError(resp, err)
+	}
+
+	simpleSpans := make([]simplespan.SimpleSpan, 0)
+	switch req.Header.Get("Content-Type") {
+	case "application/x-msgpack", "application/msgpack":
+		decoder := msgpack.NewDecoder(bytes.NewReader(body))
+		decoder.UseLooseInterfaceDecoding(true)
+		err = decoder.Decode(&simpleSpans)
+		if err != nil {
+			r.settings.Logger.Info("messagepack decoding failed")
+		}
+		if len(simpleSpans) > 0 {
+			r.settings.Logger.Debug("Decoding with msgpack worked", zap.Time("timestamp.first.msgpacktimestamp", *simpleSpans[0].MsgPackTimestamp), zap.String("timestamp.first.time", simpleSpans[0].Time))
+			r.settings.Logger.Debug("span zero", zap.String("span.data", simpleSpans[0].DebugString()))
+		}
+	case jsonContentType:
+		err = json.Unmarshal(body, &simpleSpans)
+		if err != nil {
+			errorutil.HTTPError(resp, err)
+		}
+		if len(simpleSpans) > 0 {
+			r.settings.Logger.Debug("Decoding with json worked", zap.Time("timestamp.first.msgpacktimestamp", *simpleSpans[0].MsgPackTimestamp), zap.String("timestamp.first.time", simpleSpans[0].Time))
+		}
+	}
+
+	otlpLogs := toPsomething(dataset, simpleSpans, *r.cfg, *r.settings.Logger)
+
+	numLogs := otlpLogs.LogRecordCount()
+	if numLogs > 0 {
+		ctx := r.obsreport.StartLogsOp(context.Background())
+		err = r.nextLogs.ConsumeLogs(ctx, otlpLogs)
+		r.obsreport.EndLogsOp(ctx, "protobuf", numLogs, err)
+	}
+
+	if err != nil {
+		errorutil.HTTPError(resp, err)
+		return
+	}
```

Comment:
Will this receiver support converting individual events in the incoming to request to different signals? Like an request payload could contain both spans and logs?

Reply:
If we require the entire incoming request payload to be a specific type we should do branching logic here.

If any `SimpleSpan` can be either a span or a log then it may be better to loop through the simplespan slice here and parse each span individually, building up a `plog.Logs` or `ptrace.Traces` as we go.